lab2+scanner/parser/LR.py

```diff
@@ -36,15 +44,120 @@ def canonical_collection(self):
+                                sym += p[1][j + 1]
+                                j += 1
+                        else:
+                            sym = p[1][i + 1]
+                        new_s = self.closure(self.goto(s, sym, idx))
                         if new_s not in set_of_states:
+                            set_of_gotos.append((idx, sym, len(set_of_states)))
                             set_of_states.append(new_s)
-        return set_of_states
+                        else:
+                            set_of_gotos.append((idx, sym, set_of_states.index(new_s)))
+                        break
+        return set_of_states, set_of_gotos
 
     def print_canonical_collection(self, can_col):
         for i in range(len(can_col)):
             print("s" + str(i) + ":" + str(can_col[i]))
+
+    def compute_parsing_table(self, can_col):
+        parsing_table = {}
+        list_of_symbols = self.g.E + self.g.N
+        set_of_states = can_col[0]
+        set_of_gotos = can_col[1]
+        for idx_of_s, s in enumerate(set_of_states):
+            row_for_s = {'action': ''}
+            for symbol in list_of_symbols:
+                row_for_s[symbol] = 'error'
+            for elem in s:
+                if elem == ('T', 'S.') or elem == ('T', 'P.'):
+                    row_for_s['action'] = 'acc'
+            if row_for_s['action'] != 'acc':
+                for i, t in enumerate(set_of_gotos):
+                    if t[0] == idx_of_s:
+                        row_for_s['action'] = 'shift'
+                        row_for_s[t[1]] = 's' + str(t[2])
+                for prod in s:
+                    if prod[1][len(prod[1]) - 1] == '.':
+                        prod_rhs = prod[1].replace('.', '')
+                        if prod[0] in self.g.P.keys():
+                            for p in self.g.P[prod[0]]:
+                                if p[0].replace(" ", "") == prod_rhs:
+                                    if row_for_s['action'] == 'shift':
+                                        raise ValueError("Shift reduce conflict at ", idx_of_s, p, prod[0])
+                                    elif 'reduce' in row_for_s['action']:
+                                        raise ValueError("Reduce reduce conflict at ", idx_of_s)
+                                    row_for_s['action'] = 'reduce ' + str(p[1])
+            parsing_table['s' + str(idx_of_s)] = row_for_s
+        return parsing_table
+
+    def parse_sequence(self, seq_file, parsing_table):
+        w = ''
+        with open(seq_file, 'r') as file:
+            for line in file:
+                w = line
+                break
+        work_stack = [('$', 's0')]
+        if "PIF" in seq_file:
+            input_stack = w.split()
+        else:
+            input_stack = [l for l in w]
+        output_band = []
+        tree_stack = []
+        parsing_tree = []
+        current_idx = 0
+        while input_stack or work_stack:
```

Comment:
I think it would be safer to throw an error after the while ends because in that case the sequence is not accepted. For example if the work stack ends ( we reached acc ) and the input stack still has symbols, then the program should output an error.